cluster/src/test/java/org/apache/iotdb/cluster/query/reader/RemoteSeriesReaderByTimestampTest.java

```diff
@@ -71,18 +74,93 @@ public void fetchSingleSeriesByTimestamp(Node header, long readerId, long time,
-        metaGroupMember);
+    PartitionGroup group = new PartitionGroup();
+    group.add(TestUtils.getNode(0));
+    group.add(TestUtils.getNode(1));
+    group.add(TestUtils.getNode(2));
+
+    SingleSeriesQueryRequest request = new SingleSeriesQueryRequest();
+    RemoteQueryContext context = new RemoteQueryContext(1);
+
+    DataSourceInfo sourceInfo = new DataSourceInfo(group, TSDataType.DOUBLE,
+      request, context, metaGroupMember, group);
+    sourceInfo.nextDataClient(true, Long.MIN_VALUE);
+
+    RemoteSeriesReaderByTimestamp reader = new RemoteSeriesReaderByTimestamp(sourceInfo);
+
     for (int i = 0; i < 100; i++) {
       assertEquals(i * 1.0, reader.getValueInTimestamp(i));
     }
     assertNull(reader.getValueInTimestamp(101));
   }
+
+  @Test
+  public void testFailedNode() throws IOException {
+    batchData = TestUtils.genBatchData(TSDataType.DOUBLE, 0, 100);
+    PartitionGroup group = new PartitionGroup();
+    group.add(TestUtils.getNode(0));
+    group.add(TestUtils.getNode(1));
+    group.add(TestUtils.getNode(2));
+
+    SingleSeriesQueryRequest request = new SingleSeriesQueryRequest();
+    RemoteQueryContext context = new RemoteQueryContext(1);
+
+    DataSourceInfo sourceInfo = new DataSourceInfo(group, TSDataType.DOUBLE,
+      request, context, metaGroupMember, group);
+    sourceInfo.nextDataClient(true, Long.MIN_VALUE);
+    RemoteSeriesReaderByTimestamp reader = new RemoteSeriesReaderByTimestamp(sourceInfo);
+
+    // normal read
+    assertEquals(TestUtils.getNode(0), sourceInfo.getCurrentNode());
+    for (int i = 0; i < 50; i++) {
+      assertEquals(i * 1.0, reader.getValueInTimestamp(i));
+    }
+
+    // a bad client, change to another node
+    DataClient badClient = new DataClient(null, null, TestUtils.getNode(0), null) {
+      @Override
+      public void fetchSingleSeriesByTimestamp(Node header, long readerId, long time,
+                                               AsyncMethodCallback<ByteBuffer> resultHandler) throws TException {
+        throw new TException("Good bye.");
+      }
+
+      @Override
+      public void querySingleSeriesByTimestamp(SingleSeriesQueryRequest request, AsyncMethodCallback<Long> resultHandler) throws TException {
+        resultHandler.onComplete(1L);
+      }
+    };
+
+    reader.setClientForTest(badClient);
+    for (int i = 50; i < 80; i++) {
+      assertEquals(i * 1.0, reader.getValueInTimestamp(i));
+    }
+    assertEquals(TestUtils.getNode(1), sourceInfo.getCurrentNode());
+
+    // a bad client, change to another node again
+    reader.setClientForTest(badClient);
+    for (int i = 80; i < 90; i++) {
+      assertEquals(i * 1.0, reader.getValueInTimestamp(i));
+    }
+    assertEquals(TestUtils.getNode(2), sourceInfo.getCurrentNode());
+
+    // all node failed
+    reader.setClientForTest(badClient);
+    try {
+      reader.getValueInTimestamp(90);
+      fail();
+    } catch (IOException e) {
+      //ignore
```

Comment:
It would be better to check the information or cause of the exception and see if it is what you have expected.